Framework/Core/include/Framework/AnalysisHelpers.h

```diff
@@ -143,69 +189,51 @@ struct ProducesGroup {
 };
 
 /// Helper template for table transformations
-template <typename METADATA>
+template <o2::aod::is_metadata M, soa::TableRef Ref>
 struct TableTransform {
-  using SOURCES = typename METADATA::sources;
-  using ORIGINALS = typename METADATA::originals;
-
-  using metadata = METADATA;
-  using sources = SOURCES;
-
-  constexpr auto sources_pack() const
-  {
-    return SOURCES{};
-  }
+  using metadata = M;
+  constexpr static auto sources = M::sources;
 
-  constexpr auto originals_pack() const
+  template <soa::TableRef R>
+  static constexpr auto base_spec()
   {
-    return ORIGINALS{};
+    return soa::tableRef2InputSpec<R>();
   }
 
-  template <typename Oi>
-  constexpr auto base_spec() const
+  static auto base_specs()
   {
-    using o_metadata = typename aod::MetadataTrait<Oi>::metadata;
-    return InputSpec{
-      o_metadata::tableLabel(),
-      header::DataOrigin{o_metadata::origin()},
-      header::DataDescription{o_metadata::description()},
-      o_metadata::version()};
-  }
-
-  template <typename... Os>
-  std::vector<InputSpec> base_specs_impl(framework::pack<Os...>) const
-  {
-    return {base_spec<Os>()...};
-  }
-
-  std::vector<InputSpec> base_specs() const
-  {
-    return base_specs_impl(sources_pack());
+    return []<size_t... Is>(std::index_sequence<Is...>) -> std::vector<InputSpec> {
+      return {base_spec<sources[Is]>()...};
+    }(std::make_index_sequence<sources.size()>{});
   }
 
   constexpr auto spec() const
   {
-    return OutputSpec{OutputLabel{METADATA::tableLabel()}, METADATA::origin(), METADATA::description(), METADATA::version()};
+    return soa::tableRef2OutputSpec<Ref>();
   }
 
   constexpr auto output() const
   {
-    return Output{METADATA::origin(), METADATA::description(), METADATA::version()};
+    return soa::tableRef2Output<Ref>();
   }
 
   constexpr auto ref() const
   {
-    return OutputRef{METADATA::tableLabel(), METADATA::version()};
+    return soa::tableRef2OutputRef<Ref>();
   }
 };
 
 /// This helper struct allows you to declare extended tables which should be
 /// created by the task (as opposed to those pre-defined by data model)
 template <typename T>
-struct Spawns : TableTransform<typename aod::MetadataTrait<framework::pack_head_t<typename T::originals>>::metadata> {
-  using extension_t = framework::pack_head_t<typename T::originals>;
-  using base_table_t = typename aod::MetadataTrait<extension_t>::metadata::base_table_t;
-  using expression_pack_t = typename aod::MetadataTrait<extension_t>::metadata::expression_pack_t;
+concept spawnable = soa::is_table<T> && soa::has_metadata<T>;
+
+template <spawnable T>
+struct Spawns : TableTransform<typename aod::MetadataTrait<o2::aod::Hash<T::ref.desc_hash>>::metadata, T::ref> {
+  using metadata = TableTransform<typename aod::MetadataTrait<o2::aod::Hash<T::ref.desc_hash>>::metadata, T::ref>::metadata;
+  using extension_t = typename metadata::extension_table_t;
```

Comment:
In retrospect these inline types are a bad idea, because the compiler is forced to instanciate them. Don't we have a better way of doing it?

Reply:
I needed a common base class template for both `Builds` and `Spawns`. It is possible to define a function and put here a decltype of it, rather than the type name directly.